// angular-leaflet/esm5/core/base-layer.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { ContentChildren, QueryList } from '@angular/core';
import { LayerHandler } from './layer-handler';
import { v4 as uuid } from 'uuid';
/**
 * @template T
 */
var BaseLayer = /** @class */ (function () {
    function BaseLayer() {
        this.handlerIds = [];
        this.id = uuid();
    }
    /**
     * @param {?} map
     * @return {?}
     */
    BaseLayer.prototype.addTo = /**
     * @param {?} map
     * @return {?}
     */
    function (map) {
        this.mapRef = map;
    };
    /**
     * @return {?}
     */
    BaseLayer.prototype.removeFrom = /**
     * @return {?}
     */
    function () {
        if (this.layerRef) {
            ((/** @type {?} */ (this.mapRef))).removeLayer(this.layerRef);
        }
    };
    /**
     * @return {?}
     */
    BaseLayer.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        this.initHandlers();
        this.handlerSubscription = this.handlers.changes.subscribe(this.initHandlers.bind(this));
    };
    /**
     * @return {?}
     */
    BaseLayer.prototype.initHandlers = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.handlers.filter((/**
         * @param {?} handler
         * @return {?}
         */
        function (handler) { return !_this.handlerIds.includes(handler.id); })).forEach(this.initHandler.bind(this));
    };
    /**
     * @param {?} handler
     * @return {?}
     */
    BaseLayer.prototype.initHandler = /**
     * @param {?} handler
     * @return {?}
     */
    function (handler) {
        handler.initialize(this.layerRef['_map'], this.layerRef);
        this.handlerIds = tslib_1.__spread(this.handlerIds, [handler.id]);
    };
    /**
     * @return {?}
     */
    BaseLayer.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.handlerSubscription.unsubscribe();
        this.removeFrom();
    };
    BaseLayer.propDecorators = {
        handlers: [{ type: ContentChildren, args: [LayerHandler,] }]
    };
    return BaseLayer;
}());
export { BaseLayer };
if (false) {
    /** @type {?} */
    BaseLayer.prototype.id;
    /** @type {?} */
    BaseLayer.prototype.mapRef;
    /** @type {?} */
    BaseLayer.prototype.layerRef;
    /** @type {?} */
    BaseLayer.prototype.handlerIds;
    /** @type {?} */
    BaseLayer.prototype.handlerSubscription;
    /** @type {?} */
    BaseLayer.prototype.handlers;
}
