// angular-leaflet/esm5/core/leaflet.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ContentChildren, ElementRef, Input, QueryList } from '@angular/core';
import { map } from 'leaflet';
import { BaseLayer } from './base-layer';
import { MapHandler } from './map-handler';
var LeafletComponent = /** @class */ (function () {
    function LeafletComponent(elementRef) {
        this.elementRef = elementRef;
        this.layerIds = [];
        this.handlerIds = [];
    }
    /**
     * @return {?}
     */
    LeafletComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.map = map(this.elementRef.nativeElement, tslib_1.__assign({}, this.options));
    };
    /**
     * @return {?}
     */
    LeafletComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        this.addLayers();
        this.layers.changes.subscribe(this.addLayers.bind(this));
        this.addHandlers();
        this.handlers.changes.subscribe(this.addHandlers.bind(this));
    };
    /**
     * @private
     * @return {?}
     */
    LeafletComponent.prototype.addLayers = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.layers.filter((/**
         * @param {?} layer
         * @return {?}
         */
        function (layer) { return !_this.layerIds.includes(layer.id); })).forEach(this.addLayer.bind(this));
    };
    /**
     * @private
     * @param {?} layer
     * @return {?}
     */
    LeafletComponent.prototype.addLayer = /**
     * @private
     * @param {?} layer
     * @return {?}
     */
    function (layer) {
        layer.addTo(this.map);
        this.layerIds = tslib_1.__spread(this.layerIds, [layer.id]);
    };
    /**
     * @private
     * @return {?}
     */
    LeafletComponent.prototype.addHandlers = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.handlers.filter((/**
         * @param {?} handler
         * @return {?}
         */
        function (handler) { return !_this.handlerIds.includes(handler.id); })).forEach(this.addHandler.bind(this));
    };
    /**
     * @private
     * @param {?} handler
     * @return {?}
     */
    LeafletComponent.prototype.addHandler = /**
     * @private
     * @param {?} handler
     * @return {?}
     */
    function (handler) {
        handler.initialize(this.map);
        this.handlerIds = tslib_1.__spread(this.handlerIds, [handler.id]);
    };
    LeafletComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngl-leaflet',
                    template: '<ng-content></ng-content>',
                    styles: ["\n      :host {\n        display: flex;\n        flex: 1 1 auto;\n      }\n    "]
                }] }
    ];
    /** @nocollapse */
    LeafletComponent.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    LeafletComponent.propDecorators = {
        options: [{ type: Input }],
        layers: [{ type: ContentChildren, args: [BaseLayer,] }],
        handlers: [{ type: ContentChildren, args: [MapHandler, { descendants: true },] }]
    };
    return LeafletComponent;
}());
export { LeafletComponent };
if (false) {
    /** @type {?} */
    LeafletComponent.prototype.options;
    /** @type {?} */
    LeafletComponent.prototype.layers;
    /** @type {?} */
    LeafletComponent.prototype.handlers;
    /** @type {?} */
    LeafletComponent.prototype.map;
    /** @type {?} */
    LeafletComponent.prototype.layerIds;
    /** @type {?} */
    LeafletComponent.prototype.handlerIds;
    /**
     * @type {?}
     * @private
     */
    LeafletComponent.prototype.elementRef;
}
