// angular-leaflet/esm2015/marker/marker.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef, Input } from '@angular/core';
import { divIcon, marker } from 'leaflet';
import { BaseLayer } from '../core/base-layer';
export class MarkerComponent extends BaseLayer {
    /**
     * @param {?} value
     * @return {?}
     */
    set latLng(value) {
        this._latLng = value;
        this.prepareLayer();
    }
    /**
     * @private
     * @return {?}
     */
    prepareLayer() {
        if (this.mapRef && this._latLng) {
            if (this.layerRef) {
                this.layerRef.setLatLng(this._latLng);
            }
            else {
                this.createLayer();
            }
        }
    }
    /**
     * @private
     * @return {?}
     */
    createLayer() {
        this.layerRef = marker(this._latLng, {
            icon: divIcon()
        });
        this.mapRef.addLayer(this.layerRef);
        this.initHandlers();
    }
    /**
     * @param {?} map
     * @return {?}
     */
    addTo(map) {
        super.addTo(map);
        this.prepareLayer();
    }
}
MarkerComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngl-marker',
                template: '',
                providers: [{ provide: BaseLayer, useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => MarkerComponent)) }]
            }] }
];
MarkerComponent.propDecorators = {
    latLng: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    MarkerComponent.prototype._latLng;
}
