// angular-leaflet/esm2015/vectors/polygon/polygon.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef, Input } from '@angular/core';
import { polygon } from 'leaflet';
import { BaseLayer } from '../../core/base-layer';
export class PolygonComponent extends BaseLayer {
    /**
     * @param {?} value
     * @return {?}
     */
    set coordinates(value) {
        this._coordinates = value;
        if (this.layerRef) {
            this.layerRef.setLatLngs(value);
        }
        else {
            this.createLayer();
        }
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set options(value) {
        this._options = value;
        if (this.layerRef) {
            this.layerRef.setStyle(value);
        }
        else {
            this.createLayer();
        }
    }
    /**
     * @return {?}
     */
    createLayer() {
        if (this.mapRef && this._coordinates) {
            this.layerRef = polygon(this._coordinates, Object.assign({}, this._options));
            this.mapRef.addLayer(this.layerRef);
            this.initHandlers();
        }
    }
    /**
     * @param {?} map
     * @return {?}
     */
    addTo(map) {
        super.addTo(map);
        if (!this.layerRef) {
            this.createLayer();
        }
    }
}
PolygonComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngl-polygon',
                template: '',
                providers: [{ provide: BaseLayer, useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => PolygonComponent)) }]
            }] }
];
PolygonComponent.propDecorators = {
    coordinates: [{ type: Input }],
    options: [{ type: Input }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    PolygonComponent.prototype._coordinates;
    /**
     * @type {?}
     * @private
     */
    PolygonComponent.prototype._options;
}
