// angular-leaflet/esm5/vectors/rectangle/rectangle.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef, Input } from '@angular/core';
import { LatLngBounds, rectangle } from 'leaflet';
import { BaseLayer } from '../../core/base-layer';
var RectangleComponent = /** @class */ (function (_super) {
    tslib_1.__extends(RectangleComponent, _super);
    function RectangleComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(RectangleComponent.prototype, "bounds", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._bounds = value;
            if (this.layerRef) {
                this.layerRef.setBounds(value);
            }
            else {
                this.createLayer();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(RectangleComponent.prototype, "options", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._options = value;
            if (this.layerRef) {
                this.layerRef.setStyle(value);
            }
            else {
                this.createLayer();
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    RectangleComponent.prototype.createLayer = /**
     * @return {?}
     */
    function () {
        if (this.mapRef && this._bounds) {
            this.layerRef = rectangle(this._bounds, tslib_1.__assign({}, this._options));
            this.mapRef.addLayer(this.layerRef);
            this.initHandlers();
        }
    };
    /**
     * @param {?} map
     * @return {?}
     */
    RectangleComponent.prototype.addTo = /**
     * @param {?} map
     * @return {?}
     */
    function (map) {
        _super.prototype.addTo.call(this, map);
        if (!this.layerRef) {
            this.createLayer();
        }
    };
    RectangleComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngl-rectangle',
                    template: '',
                    providers: [{ provide: BaseLayer, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return RectangleComponent; })) }]
                }] }
    ];
    RectangleComponent.propDecorators = {
        bounds: [{ type: Input }],
        options: [{ type: Input }]
    };
    return RectangleComponent;
}(BaseLayer));
export { RectangleComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    RectangleComponent.prototype._bounds;
    /**
     * @type {?}
     * @private
     */
    RectangleComponent.prototype._options;
}
