// angular-leaflet/esm2015/tile-layer/tile-layer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef } from '@angular/core';
import { tileLayer } from 'leaflet';
import { BaseLayer } from '../core/base-layer';
export class TileLayerComponent extends BaseLayer {
    /**
     * @param {?} map
     * @return {?}
     */
    addTo(map) {
        super.addTo(map);
        if (!this.layerRef) {
            this.layerRef = tileLayer('http://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png');
            this.mapRef.addLayer(this.layerRef);
            this.initHandlers();
        }
    }
}
TileLayerComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngl-tile-layer',
                template: '',
                providers: [{ provide: BaseLayer, useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => TileLayerComponent)) }]
            }] }
];
