// angular-leaflet/esm5/vectors/circle/circle.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef, Input } from '@angular/core';
import { BaseLayer } from '../../core/base-layer';
import { circle } from 'leaflet';
var CircleComponent = /** @class */ (function (_super) {
    tslib_1.__extends(CircleComponent, _super);
    function CircleComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(CircleComponent.prototype, "center", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._center = value;
            if (this.layerRef) {
                this.layerRef.setLatLng(value);
            }
            else {
                this.createLayer();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CircleComponent.prototype, "radius", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._radius = value;
            if (this.layerRef) {
                this.layerRef.setRadius(value);
            }
            else {
                this.createLayer();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CircleComponent.prototype, "options", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._options = value;
            if (this.layerRef) {
                this.layerRef.setStyle(value);
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} map
     * @return {?}
     */
    CircleComponent.prototype.addTo = /**
     * @param {?} map
     * @return {?}
     */
    function (map) {
        _super.prototype.addTo.call(this, map);
        if (!this.layerRef) {
            this.createLayer();
        }
    };
    /**
     * @private
     * @return {?}
     */
    CircleComponent.prototype.createLayer = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.mapRef && this._center && this._radius) {
            this.layerRef = circle(this._center, tslib_1.__assign({}, this._options, { radius: this._radius }));
            this.mapRef.addLayer(this.layerRef);
            this.initHandlers();
        }
    };
    CircleComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngl-circle',
                    template: '',
                    providers: [{ provide: BaseLayer, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return CircleComponent; })) }]
                }] }
    ];
    CircleComponent.propDecorators = {
        center: [{ type: Input }],
        radius: [{ type: Input }],
        options: [{ type: Input }]
    };
    return CircleComponent;
}(BaseLayer));
export { CircleComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    CircleComponent.prototype._center;
    /**
     * @type {?}
     * @private
     */
    CircleComponent.prototype._radius;
    /**
     * @type {?}
     * @private
     */
    CircleComponent.prototype._options;
}
