// angular-leaflet/esm5/vectors/polyline/polyline.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef, Input } from '@angular/core';
import { polyline } from 'leaflet';
import { BaseLayer } from '../../core/base-layer';
var PolylineComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PolylineComponent, _super);
    function PolylineComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(PolylineComponent.prototype, "coordinates", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._coordinates = value;
            if (this.layerRef) {
                this.layerRef.setLatLngs(value);
            }
            else {
                this.createLayer();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PolylineComponent.prototype, "options", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._options = value;
            if (this.layerRef) {
                this.layerRef.setStyle(value);
            }
            else {
                this.createLayer();
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} map
     * @return {?}
     */
    PolylineComponent.prototype.addTo = /**
     * @param {?} map
     * @return {?}
     */
    function (map) {
        _super.prototype.addTo.call(this, map);
        if (!this.layerRef && this._coordinates) {
            this.createLayer();
        }
    };
    /**
     * @private
     * @return {?}
     */
    PolylineComponent.prototype.createLayer = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.mapRef && this._coordinates) {
            this.layerRef = polyline(this._coordinates, tslib_1.__assign({}, this._options));
            this.mapRef.addLayer(this.layerRef);
            this.initHandlers();
        }
    };
    PolylineComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngl-polyline',
                    template: '',
                    providers: [{ provide: BaseLayer, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return PolylineComponent; })) }]
                }] }
    ];
    PolylineComponent.propDecorators = {
        coordinates: [{ type: Input }],
        options: [{ type: Input }]
    };
    return PolylineComponent;
}(BaseLayer));
export { PolylineComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PolylineComponent.prototype._coordinates;
    /**
     * @type {?}
     * @private
     */
    PolylineComponent.prototype._options;
}
