// angular-leaflet/esm5/marker/marker.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef, Input } from '@angular/core';
import { divIcon, marker } from 'leaflet';
import { BaseLayer } from '../core/base-layer';
var MarkerComponent = /** @class */ (function (_super) {
    tslib_1.__extends(MarkerComponent, _super);
    function MarkerComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(MarkerComponent.prototype, "latLng", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._latLng = value;
            this.prepareLayer();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @private
     * @return {?}
     */
    MarkerComponent.prototype.prepareLayer = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.mapRef && this._latLng) {
            if (this.layerRef) {
                this.layerRef.setLatLng(this._latLng);
            }
            else {
                this.createLayer();
            }
        }
    };
    /**
     * @private
     * @return {?}
     */
    MarkerComponent.prototype.createLayer = /**
     * @private
     * @return {?}
     */
    function () {
        this.layerRef = marker(this._latLng, {
            icon: divIcon()
        });
        this.mapRef.addLayer(this.layerRef);
        this.initHandlers();
    };
    /**
     * @param {?} map
     * @return {?}
     */
    MarkerComponent.prototype.addTo = /**
     * @param {?} map
     * @return {?}
     */
    function (map) {
        _super.prototype.addTo.call(this, map);
        this.prepareLayer();
    };
    MarkerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngl-marker',
                    template: '',
                    providers: [{ provide: BaseLayer, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return MarkerComponent; })) }]
                }] }
    ];
    MarkerComponent.propDecorators = {
        latLng: [{ type: Input }]
    };
    return MarkerComponent;
}(BaseLayer));
export { MarkerComponent };
if (false) {
    /** @type {?} */
    MarkerComponent.prototype._latLng;
}
