// angular-leaflet/esm5/image-overlay/image-overlay.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef, Input } from '@angular/core';
import { imageOverlay, latLngBounds } from 'leaflet';
import { BaseLayer } from '../core/base-layer';
var ImageOverlayComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ImageOverlayComponent, _super);
    function ImageOverlayComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(ImageOverlayComponent.prototype, "src", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._url = value;
            if (this.layerRef) {
                this.layerRef.setUrl(value);
            }
            else {
                this.prepareLayer();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ImageOverlayComponent.prototype, "width", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._width = value;
            if (this.layerRef) {
                this.layerRef.setBounds(this.prepareBounds(this._width, this._height));
            }
            else {
                this.prepareLayer();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ImageOverlayComponent.prototype, "height", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._height = value;
            if (this.layerRef) {
                this.layerRef.setBounds(this.prepareBounds(this._width, this._height));
            }
            else {
                this.prepareLayer();
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} map
     * @return {?}
     */
    ImageOverlayComponent.prototype.addTo = /**
     * @param {?} map
     * @return {?}
     */
    function (map) {
        _super.prototype.addTo.call(this, map);
        this.prepareLayer();
    };
    /**
     * @return {?}
     */
    ImageOverlayComponent.prototype.prepareLayer = /**
     * @return {?}
     */
    function () {
        if (this.mapRef && this._url && this._height && this._width) {
            this.layerRef = imageOverlay(this._url, this.prepareBounds(this._width, this._height));
            this.mapRef.addLayer(this.layerRef);
            this.initHandlers();
        }
    };
    /**
     * @private
     * @param {?} width
     * @param {?} height
     * @return {?}
     */
    ImageOverlayComponent.prototype.prepareBounds = /**
     * @private
     * @param {?} width
     * @param {?} height
     * @return {?}
     */
    function (width, height) {
        /** @type {?} */
        var southWest = ((/** @type {?} */ (this.mapRef))).unproject([0, height * 2], 1);
        /** @type {?} */
        var northEast = ((/** @type {?} */ (this.mapRef))).unproject([width * 2, 0], 1);
        return latLngBounds(southWest, northEast);
    };
    ImageOverlayComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngl-image-overlay',
                    template: '',
                    providers: [{ provide: BaseLayer, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return ImageOverlayComponent; })) }]
                }] }
    ];
    ImageOverlayComponent.propDecorators = {
        src: [{ type: Input }],
        width: [{ type: Input }],
        height: [{ type: Input }]
    };
    return ImageOverlayComponent;
}(BaseLayer));
export { ImageOverlayComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ImageOverlayComponent.prototype._url;
    /**
     * @type {?}
     * @private
     */
    ImageOverlayComponent.prototype._width;
    /**
     * @type {?}
     * @private
     */
    ImageOverlayComponent.prototype._height;
}
