// angular-leaflet/esm2015/vectors/circle/circle.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef, Input } from '@angular/core';
import { BaseLayer } from '../../core/base-layer';
import { circle } from 'leaflet';
export class CircleComponent extends BaseLayer {
    /**
     * @param {?} value
     * @return {?}
     */
    set center(value) {
        this._center = value;
        if (this.layerRef) {
            this.layerRef.setLatLng(value);
        }
        else {
            this.createLayer();
        }
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set radius(value) {
        this._radius = value;
        if (this.layerRef) {
            this.layerRef.setRadius(value);
        }
        else {
            this.createLayer();
        }
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set options(value) {
        this._options = value;
        if (this.layerRef) {
            this.layerRef.setStyle(value);
        }
    }
    /**
     * @param {?} map
     * @return {?}
     */
    addTo(map) {
        super.addTo(map);
        if (!this.layerRef) {
            this.createLayer();
        }
    }
    /**
     * @private
     * @return {?}
     */
    createLayer() {
        if (this.mapRef && this._center && this._radius) {
            this.layerRef = circle(this._center, Object.assign({}, this._options, { radius: this._radius }));
            this.mapRef.addLayer(this.layerRef);
            this.initHandlers();
        }
    }
}
CircleComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngl-circle',
                template: '',
                providers: [{ provide: BaseLayer, useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => CircleComponent)) }]
            }] }
];
CircleComponent.propDecorators = {
    center: [{ type: Input }],
    radius: [{ type: Input }],
    options: [{ type: Input }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    CircleComponent.prototype._center;
    /**
     * @type {?}
     * @private
     */
    CircleComponent.prototype._radius;
    /**
     * @type {?}
     * @private
     */
    CircleComponent.prototype._options;
}
