// angular-leaflet/esm2015/vectors/rectangle/rectangle.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef, Input } from '@angular/core';
import { LatLngBounds, rectangle } from 'leaflet';
import { BaseLayer } from '../../core/base-layer';
export class RectangleComponent extends BaseLayer {
    /**
     * @param {?} value
     * @return {?}
     */
    set bounds(value) {
        this._bounds = value;
        if (this.layerRef) {
            this.layerRef.setBounds(value);
        }
        else {
            this.createLayer();
        }
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set options(value) {
        this._options = value;
        if (this.layerRef) {
            this.layerRef.setStyle(value);
        }
        else {
            this.createLayer();
        }
    }
    /**
     * @return {?}
     */
    createLayer() {
        if (this.mapRef && this._bounds) {
            this.layerRef = rectangle(this._bounds, Object.assign({}, this._options));
            this.mapRef.addLayer(this.layerRef);
            this.initHandlers();
        }
    }
    /**
     * @param {?} map
     * @return {?}
     */
    addTo(map) {
        super.addTo(map);
        if (!this.layerRef) {
            this.createLayer();
        }
    }
}
RectangleComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngl-rectangle',
                template: '',
                providers: [{ provide: BaseLayer, useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => RectangleComponent)) }]
            }] }
];
RectangleComponent.propDecorators = {
    bounds: [{ type: Input }],
    options: [{ type: Input }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    RectangleComponent.prototype._bounds;
    /**
     * @type {?}
     * @private
     */
    RectangleComponent.prototype._options;
}
