// angular-leaflet/esm2015/image-overlay/image-overlay.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, forwardRef, Input } from '@angular/core';
import { imageOverlay, latLngBounds } from 'leaflet';
import { BaseLayer } from '../core/base-layer';
export class ImageOverlayComponent extends BaseLayer {
    /**
     * @param {?} value
     * @return {?}
     */
    set src(value) {
        this._url = value;
        if (this.layerRef) {
            this.layerRef.setUrl(value);
        }
        else {
            this.prepareLayer();
        }
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set width(value) {
        this._width = value;
        if (this.layerRef) {
            this.layerRef.setBounds(this.prepareBounds(this._width, this._height));
        }
        else {
            this.prepareLayer();
        }
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set height(value) {
        this._height = value;
        if (this.layerRef) {
            this.layerRef.setBounds(this.prepareBounds(this._width, this._height));
        }
        else {
            this.prepareLayer();
        }
    }
    /**
     * @param {?} map
     * @return {?}
     */
    addTo(map) {
        super.addTo(map);
        this.prepareLayer();
    }
    /**
     * @return {?}
     */
    prepareLayer() {
        if (this.mapRef && this._url && this._height && this._width) {
            this.layerRef = imageOverlay(this._url, this.prepareBounds(this._width, this._height));
            this.mapRef.addLayer(this.layerRef);
            this.initHandlers();
        }
    }
    /**
     * @private
     * @param {?} width
     * @param {?} height
     * @return {?}
     */
    prepareBounds(width, height) {
        /** @type {?} */
        const southWest = ((/** @type {?} */ (this.mapRef))).unproject([0, height * 2], 1);
        /** @type {?} */
        const northEast = ((/** @type {?} */ (this.mapRef))).unproject([width * 2, 0], 1);
        return latLngBounds(southWest, northEast);
    }
}
ImageOverlayComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngl-image-overlay',
                template: '',
                providers: [{ provide: BaseLayer, useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => ImageOverlayComponent)) }]
            }] }
];
ImageOverlayComponent.propDecorators = {
    src: [{ type: Input }],
    width: [{ type: Input }],
    height: [{ type: Input }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    ImageOverlayComponent.prototype._url;
    /**
     * @type {?}
     * @private
     */
    ImageOverlayComponent.prototype._width;
    /**
     * @type {?}
     * @private
     */
    ImageOverlayComponent.prototype._height;
}
