// angular-leaflet/esm5/groups/feature-group/feature-group.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ContentChildren, forwardRef, QueryList } from '@angular/core';
import { BaseLayer } from '../../core/base-layer';
import { FeatureGroup } from 'leaflet';
var FeatureGroupComponent = /** @class */ (function (_super) {
    tslib_1.__extends(FeatureGroupComponent, _super);
    function FeatureGroupComponent() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.layerIds = [];
        return _this;
    }
    /**
     * @param {?} map
     * @return {?}
     */
    FeatureGroupComponent.prototype.addTo = /**
     * @param {?} map
     * @return {?}
     */
    function (map) {
        _super.prototype.addTo.call(this, map);
        this.layerRef = new FeatureGroup();
        this.mapRef.addLayer(this.layerRef);
        this.addLayers();
        this.initHandlers();
        this.layers.changes.subscribe(this.addLayers.bind(this));
    };
    /**
     * @private
     * @return {?}
     */
    FeatureGroupComponent.prototype.addLayers = /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        this.layers
            .filter((/**
         * @param {?} layer
         * @return {?}
         */
        function (layer) { return layer.id !== _this.id; }))
            .filter((/**
         * @param {?} layer
         * @return {?}
         */
        function (layer) { return !_this.layerIds.includes(layer.id); }))
            .forEach(this.addLayer.bind(this));
    };
    /**
     * @private
     * @param {?} layer
     * @return {?}
     */
    FeatureGroupComponent.prototype.addLayer = /**
     * @private
     * @param {?} layer
     * @return {?}
     */
    function (layer) {
        layer.addTo(this.layerRef);
        this.layerIds = tslib_1.__spread(this.layerIds, [layer.id]);
    };
    FeatureGroupComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngl-feature-group',
                    template: '',
                    providers: [{ provide: BaseLayer, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return FeatureGroupComponent; })) }]
                }] }
    ];
    FeatureGroupComponent.propDecorators = {
        layers: [{ type: ContentChildren, args: [BaseLayer,] }]
    };
    return FeatureGroupComponent;
}(BaseLayer));
export { FeatureGroupComponent };
if (false) {
    /** @type {?} */
    FeatureGroupComponent.prototype.layers;
    /** @type {?} */
    FeatureGroupComponent.prototype.layerIds;
}
