// angular-leaflet/esm5/tile-layer/tile-layer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef } from '@angular/core';
import { tileLayer } from 'leaflet';
import { BaseLayer } from '../core/base-layer';
var TileLayerComponent = /** @class */ (function (_super) {
    tslib_1.__extends(TileLayerComponent, _super);
    function TileLayerComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @param {?} map
     * @return {?}
     */
    TileLayerComponent.prototype.addTo = /**
     * @param {?} map
     * @return {?}
     */
    function (map) {
        _super.prototype.addTo.call(this, map);
        if (!this.layerRef) {
            this.layerRef = tileLayer('http://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png');
            this.mapRef.addLayer(this.layerRef);
            this.initHandlers();
        }
    };
    TileLayerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngl-tile-layer',
                    template: '',
                    providers: [{ provide: BaseLayer, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return TileLayerComponent; })) }]
                }] }
    ];
    return TileLayerComponent;
}(BaseLayer));
export { TileLayerComponent };
