// angular-leaflet/esm5/vectors/polygon/polygon.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, forwardRef, Input } from '@angular/core';
import { polygon } from 'leaflet';
import { BaseLayer } from '../../core/base-layer';
var PolygonComponent = /** @class */ (function (_super) {
    tslib_1.__extends(PolygonComponent, _super);
    function PolygonComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(PolygonComponent.prototype, "coordinates", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._coordinates = value;
            if (this.layerRef) {
                this.layerRef.setLatLngs(value);
            }
            else {
                this.createLayer();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PolygonComponent.prototype, "options", {
        set: /**
         * @param {?} value
         * @return {?}
         */
        function (value) {
            this._options = value;
            if (this.layerRef) {
                this.layerRef.setStyle(value);
            }
            else {
                this.createLayer();
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    PolygonComponent.prototype.createLayer = /**
     * @return {?}
     */
    function () {
        if (this.mapRef && this._coordinates) {
            this.layerRef = polygon(this._coordinates, tslib_1.__assign({}, this._options));
            this.mapRef.addLayer(this.layerRef);
            this.initHandlers();
        }
    };
    /**
     * @param {?} map
     * @return {?}
     */
    PolygonComponent.prototype.addTo = /**
     * @param {?} map
     * @return {?}
     */
    function (map) {
        _super.prototype.addTo.call(this, map);
        if (!this.layerRef) {
            this.createLayer();
        }
    };
    PolygonComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ngl-polygon',
                    template: '',
                    providers: [{ provide: BaseLayer, useExisting: forwardRef((/**
                             * @return {?}
                             */
                            function () { return PolygonComponent; })) }]
                }] }
    ];
    PolygonComponent.propDecorators = {
        coordinates: [{ type: Input }],
        options: [{ type: Input }]
    };
    return PolygonComponent;
}(BaseLayer));
export { PolygonComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    PolygonComponent.prototype._coordinates;
    /**
     * @type {?}
     * @private
     */
    PolygonComponent.prototype._options;
}
