// angular-leaflet/esm2015/core/base-layer.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ContentChildren, QueryList } from '@angular/core';
import { LayerHandler } from './layer-handler';
import { v4 as uuid } from 'uuid';
/**
 * @template T
 */
export class BaseLayer {
    constructor() {
        this.handlerIds = [];
        this.id = uuid();
    }
    /**
     * @param {?} map
     * @return {?}
     */
    addTo(map) {
        this.mapRef = map;
    }
    /**
     * @return {?}
     */
    removeFrom() {
        if (this.layerRef) {
            ((/** @type {?} */ (this.mapRef))).removeLayer(this.layerRef);
        }
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this.initHandlers();
        this.handlerSubscription = this.handlers.changes.subscribe(this.initHandlers.bind(this));
    }
    /**
     * @return {?}
     */
    initHandlers() {
        this.handlers.filter((/**
         * @param {?} handler
         * @return {?}
         */
        handler => !this.handlerIds.includes(handler.id))).forEach(this.initHandler.bind(this));
    }
    /**
     * @param {?} handler
     * @return {?}
     */
    initHandler(handler) {
        handler.initialize(this.layerRef['_map'], this.layerRef);
        this.handlerIds = [...this.handlerIds, handler.id];
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.handlerSubscription.unsubscribe();
        this.removeFrom();
    }
}
BaseLayer.propDecorators = {
    handlers: [{ type: ContentChildren, args: [LayerHandler,] }]
};
if (false) {
    /** @type {?} */
    BaseLayer.prototype.id;
    /** @type {?} */
    BaseLayer.prototype.mapRef;
    /** @type {?} */
    BaseLayer.prototype.layerRef;
    /** @type {?} */
    BaseLayer.prototype.handlerIds;
    /** @type {?} */
    BaseLayer.prototype.handlerSubscription;
    /** @type {?} */
    BaseLayer.prototype.handlers;
}
