// angular-leaflet/esm2015/core/leaflet.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChildren, ElementRef, Input, QueryList } from '@angular/core';
import { map } from 'leaflet';
import { BaseLayer } from './base-layer';
import { MapHandler } from './map-handler';
export class LeafletComponent {
    /**
     * @param {?} elementRef
     */
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.layerIds = [];
        this.handlerIds = [];
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.map = map(this.elementRef.nativeElement, Object.assign({}, this.options));
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this.addLayers();
        this.layers.changes.subscribe(this.addLayers.bind(this));
        this.addHandlers();
        this.handlers.changes.subscribe(this.addHandlers.bind(this));
    }
    /**
     * @private
     * @return {?}
     */
    addLayers() {
        this.layers.filter((/**
         * @param {?} layer
         * @return {?}
         */
        layer => !this.layerIds.includes(layer.id))).forEach(this.addLayer.bind(this));
    }
    /**
     * @private
     * @param {?} layer
     * @return {?}
     */
    addLayer(layer) {
        layer.addTo(this.map);
        this.layerIds = [...this.layerIds, layer.id];
    }
    /**
     * @private
     * @return {?}
     */
    addHandlers() {
        this.handlers.filter((/**
         * @param {?} handler
         * @return {?}
         */
        handler => !this.handlerIds.includes(handler.id))).forEach(this.addHandler.bind(this));
    }
    /**
     * @private
     * @param {?} handler
     * @return {?}
     */
    addHandler(handler) {
        handler.initialize(this.map);
        this.handlerIds = [...this.handlerIds, handler.id];
    }
}
LeafletComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngl-leaflet',
                template: '<ng-content></ng-content>',
                styles: [`
      :host {
        display: flex;
        flex: 1 1 auto;
      }
    `]
            }] }
];
/** @nocollapse */
LeafletComponent.ctorParameters = () => [
    { type: ElementRef }
];
LeafletComponent.propDecorators = {
    options: [{ type: Input }],
    layers: [{ type: ContentChildren, args: [BaseLayer,] }],
    handlers: [{ type: ContentChildren, args: [MapHandler, { descendants: true },] }]
};
if (false) {
    /** @type {?} */
    LeafletComponent.prototype.options;
    /** @type {?} */
    LeafletComponent.prototype.layers;
    /** @type {?} */
    LeafletComponent.prototype.handlers;
    /** @type {?} */
    LeafletComponent.prototype.map;
    /** @type {?} */
    LeafletComponent.prototype.layerIds;
    /** @type {?} */
    LeafletComponent.prototype.handlerIds;
    /**
     * @type {?}
     * @private
     */
    LeafletComponent.prototype.elementRef;
}
