// angular-leaflet/esm2015/groups/feature-group/feature-group.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ContentChildren, forwardRef, QueryList } from '@angular/core';
import { BaseLayer } from '../../core/base-layer';
import { FeatureGroup } from 'leaflet';
export class FeatureGroupComponent extends BaseLayer {
    constructor() {
        super(...arguments);
        this.layerIds = [];
    }
    /**
     * @param {?} map
     * @return {?}
     */
    addTo(map) {
        super.addTo(map);
        this.layerRef = new FeatureGroup();
        this.mapRef.addLayer(this.layerRef);
        this.addLayers();
        this.initHandlers();
        this.layers.changes.subscribe(this.addLayers.bind(this));
    }
    /**
     * @private
     * @return {?}
     */
    addLayers() {
        this.layers
            .filter((/**
         * @param {?} layer
         * @return {?}
         */
        layer => layer.id !== this.id))
            .filter((/**
         * @param {?} layer
         * @return {?}
         */
        layer => !this.layerIds.includes(layer.id)))
            .forEach(this.addLayer.bind(this));
    }
    /**
     * @private
     * @param {?} layer
     * @return {?}
     */
    addLayer(layer) {
        layer.addTo(this.layerRef);
        this.layerIds = [...this.layerIds, layer.id];
    }
}
FeatureGroupComponent.decorators = [
    { type: Component, args: [{
                selector: 'ngl-feature-group',
                template: '',
                providers: [{ provide: BaseLayer, useExisting: forwardRef((/**
                         * @return {?}
                         */
                        () => FeatureGroupComponent)) }]
            }] }
];
FeatureGroupComponent.propDecorators = {
    layers: [{ type: ContentChildren, args: [BaseLayer,] }]
};
if (false) {
    /** @type {?} */
    FeatureGroupComponent.prototype.layers;
    /** @type {?} */
    FeatureGroupComponent.prototype.layerIds;
}
